python/cuml/dask/common/dask_arr_utils.py

```diff
@@ -160,49 +109,51 @@ def to_sp_dask_array(cudf_or_array, client=None):
 
     meta = cupyx.scipy.sparse.csr_matrix(rmm_cupy_ary(cp.zeros, 1))
 
-    if isinstance(cudf_or_array, dask.array.Array):
+    ret = cudf_or_array
+
+    if isinstance(ret, dask.array.Array):
         # At the time of developing this, using map_blocks will not work
         # to convert a Dask.Array to CuPy sparse arrays underneath.
 
-        parts = client.sync(_extract_partitions, cudf_or_array)
-        cudf_or_array = [client.submit(_conv_np_to_df, part, workers=[w])
-                         for w, part in parts]
+        parts = client.sync(_extract_partitions, ret)
+        futures = [client.submit(_conv_np_to_df, part, workers=[w], pure=False)
+                   for w, part in parts]
 
-        cudf_or_array = to_dask_cudf(cudf_or_array)
+        ret = to_dask_cudf(futures)
```

Comment:
does this increase memory usage because cudf_or_array can no longer be GC'ed? Actually, I guess the caller has a reference so we can't gc anyways...

Reply:
Unfortunately, one of the issues that seems to be triggering the race condition happens when a variable containing futures gets overwritten by something that does not. The ideal case here would be that the data in the futures is not copied at all. This function is ugly, but it's only meant to allow us to build Sparse cupy-backed Dask arrays (specifically for testing and simple examples) until we are able to do so directly.